Add new flag masterkeyfile

aawsome opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

Add the new flag `--masterkeyfile` which allows to directly use the
master key from a file.
When `--masterkeyfile` is used from `restic init` the master key is
written to the given file and no key is created in the repo.
For all other commands, `--masterkeyfile` uses the master key in the
given file. No password is needed then.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Was discussed (don't remember the issue) and is able to solve #2128 
Also allows increased security if you do want to save the password-encrypted masterkey.
On low-end machines saving the overhead from the KDF may give you some benefit.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Commit:
Add new flag masterkeyfile

Add the new flag `--masterkeyfile` which allows to directly use the
master key from a file.
When `--masterkeyfile` is used from `restic init` the master key is
written to the given file and no key is created in the repo.
For all other commands, `--masterkeyfile` uses the master key in the
given file. No password is needed.

Commit:
Add changelog and update docu

shtratos commented on cmd/restic/cmd_init.go:
```diff
@@ -33,25 +33,32 @@ func runInit(gopts GlobalOptions, args []string) error {
+		Verbosef("the repository. Losing your password means that your data is\n")
+		Verbosef("irrecoverably lost.\n")
+	} else {
+		Verbosef("Please note that you need the masterkey %s to access the repository\n", gopts.MasterKeyFile)
```
[nit] Dot is missing at the end of the line.

aawsome replied:
done.

shtratos commented on cmd/restic/global.go:
```diff
@@ -366,26 +368,38 @@ func OpenRepository(opts GlobalOptions) (*repository.Repository, error) {
+				opts.password = ""
+				fmt.Printf("%s. Try again\n", err)
+			}
 		}
```
this whole block is rather long, please consider extracting it into a separate function or method.

aawsome replied:
Thanks for the suggestion! I also changed it.

shtratos commented on doc/070_encryption.rst:
```diff
@@ -49,3 +49,51 @@ per repository. In fact, you can use the ``list``, ``add``, ``remove``, and
+    irrecoverably lost.
+    $ restic -r  /srv/restic-repo --quiet cat masterkey > key.master 
+    enter password for repository:
+    $ chmod 400 key.master
```
[paranoia] I'd suggest first creating the `key.master` file, chmodding it to `600`, then exporting the key and chmodding it to `400`.
That way it can not be read by other users during the time between the export and `chmod 400`.

aawsome replied:
Changed it to `restic ... cat ... && chmod`.

shtratos commented on internal/repository/repository.go:
```diff
@@ -568,6 +569,33 @@ func LoadIndex(ctx context.Context, repo restic.Repository, id restic.ID) (*Inde
+// MasterKeyFile reads the master key from a given file
+func (r *Repository) MasterKeyFile(ctx context.Context, keyfile string) error {
+
+	buf, err := ioutil.ReadFile(keyfile) // just pass the file name
```
[paranoia] should we check `opts.MasterKeyFile` mode and refuse to accept it unless it's `600` or `400`  like SSH does?
Maybe also warn the user that the key might have been compromised?

shtratos replied:
[nit] not sure what is the value of the `// just pass the file name` comment. Is it needed?

aawsome replied:
Not really sure if we should do paranoia here.. Do others have an opinion about it? If more are voting for a permission check, I'll add it.

Comment is removed - was a copy/paste error

shtratos commented on doc/070_encryption.rst:
```diff
@@ -49,3 +49,51 @@ per repository. In fact, you can use the ``list``, ``add``, ``remove``, and
+    $ restic -r /srv/restic-repo --masterkeyfile master.key init
+    created restic repository f855d38126 at /tmp/repo
+
+    Please note that you need the masterkey tmp.master to access the repository
```
[typo?] Did you mean `master.key` instead of `tmp.master` here?

aawsome replied:
is now corrected. Thanks!

shtratos commented on doc/070_encryption.rst:
```diff
@@ -49,3 +49,51 @@ per repository. In fact, you can use the ``list``, ``add``, ``remove``, and
+    ---------------------------------------------------
+    335ca24b  username    kasimir   2015-08-12 13:35:05
+    ---------------------------------------------------
+    $ restic -r /srv/restic-repo --masterkeyfile key.master key remove 335ca24b
```
should removing a master key from the repository warn a user that the repository can't be accessed without the master key file?

aawsome replied:
I added a general warning when deleting repo keys using masterkey file.

shtratos reviewed (commented):
Thanks for having a stab at this! 

I'm a bit confused on what is the exact value of having the masterkey file stored outside the repository.
Could you please clarify it in the documentation?

As a restic user, it's hard for me to make a judgement which option is "better" or "more secure".
Is there a tradeoff? What is it? Which one is the recommended option?
Is the masterkey file encrypted?

Commit:
Changes from to code-review

- fix typos
- add warning when deleting repository keys using masterkey file
- small refactoring of `global.go`
- add pros and cons of using masterkey file vs. repository keys to docu
-

fd0 commented:
Thank you very much for your contribution!

I'm going to be honest here: in my opinion, the complexity added by this feature outweights its usefulness. And #2128 can be addressed in different ways.

I couldn't find any discussion about this. Do you still remember where it was discussed?

aawsome commented:
Thanks @fd0 for your feedback!

As written I do not really remember where the discussion was and having searched for it I couldn't find it. I recall there was some about the possibility to just use a key instead of password but maybe my memory was misleading me.
For me this PR is absolutely useful. As for backup security is a prio one matter, I do think we should try to offer a variety of options to satisfy the user's need of security. I agree that we should think about options which enable the user to drop security. IMO this PR allows to raise security when considering some threat options like possibility of reading repos and a weakness in the KDF allowing brute-force attacks.

Personally when starting with restic I was puzzled that there was no option to just use a key as most other backups tools do offer. As a result I only use a generated password-file with enough entropy for my one-key repository. Adding the masterkeyfile option would make my setting even more secure.

About the complexity of the PR:
I do not think that adding the masterkeyfile option does add a significant complexity to the code. Can you specify which parts of the PR seems to be too complex for you?
From my point of view the majority of the changes are due to documentation, hints for users and the ability to create a masterkeyfile during repo init. I consider all of this optional and can remove it if you like. I'm also open to suggestions how to lower complexity in the few remaining parts.
Thank you!

rawtaz commented:
@aawsome So put in simple terms, the use case would be to sidestep restic's own encryption and directly provide a master key such that you can use your own encryption algorithms, for the purpose of when you don't trust the encryption restic uses?

aawsome commented:
> @aawsome So put in simple terms, the use case would be to sidestep restic's own encryption and directly provide a master key such that you can use your own encryption algorithms, for the purpose of when you don't trust the encryption restic uses?

No, this does not describe the use case correctly.

When talking about encryption of a backup tool, I would consider the encryption used to encrypt the data backed up and the cryptography used for authentication. Restic's encryption hence is AES-256 and the authentication algorithm is POLY-1305; this is not changed by own encryption/authentication when using this PR.

However when talking about the keys used for the AES encryption (and for the POLY-1305 authentication) comes the use case for this PR. Restic saves these keys in the repository using passwords and a key derivation function (KDF). Moreover it enables users to save these keys a multiple times with different passwords. There are use cases for this way of handling the encryption keys. However from a security point of view this just adds more components to the security chain. Your data is only as secure as the weakest link in the chain: AES security; KDF security; security of weakest password used.

Using this PR does not change anything about encryption or authentication of the data. It just offers the users the possibility to only rely on the security of AES-256 and POLY-1305 for the safety and authenticity of the data.

rawtaz commented:
> Using this PR does not change anything about encryption or authentication of the data. It just offers the users the possibility to only rely on the security of AES-256 and POLY-1305 for the safety and authenticity of the data.

But as you wrote, this is already what they are doing. Plus the KDF of course. Is your PR effectively just giving the option to skip the KDF part (including multiple passwords)?

aawsome commented:
> Is your PR effectively just giving the option to skip the KDF part (including multiple passwords)?

Exactly. This is what I also added to the documentation in 070_encryption.rst. Can I add/change something there to be more clear about what `--masterkeyfile` is doing?

rawtaz commented:
> Personally when starting with restic I was puzzled that there was no option to just use a key as most other backups tools do offer.

Can you give examples of those other backup tools and what arguments they provide to do what this PR is about?

Honestly, I fail to see what the relevant use for this feature is. You keep talking about raising security, but I'm not convinced since you're not being very clear about what actual concrete problems this solves. I wish you would have discussed this in an issue before you started implementing it. Issue first, PR second, anything else is backwards :)

aawsome commented:
> Can you give examples of those other backup tools and what arguments they provide to do what this PR is about?

Just consider borg. There is the option `repokey` which saves the key in the repository. This is similar to what restic is doing at the moment. And there is the option `keyfile`which allows to save the key locally. See details [here](https://borgbackup.readthedocs.io/en/stable/internals/data-structures.html#key-files)

I tried to add the option to save the key locally with minimal added complexity and thus came to this PR.

fd0 commented:
I'll clarify what I meant with "complexity": At the moment, there's only one location for the key files (the repo) and only one code path used for accessing data in the repo: get the key file from the repo, then call the kdf with the password to get the master keys.

If we add this option, there's a second code path. Restic would need to add more checks, for example that the master keys are valid and have indeed been generated with restic, the key material is long enough and so on. I can imagine all sorts of issues which could arise from having a "shortcut" around the KDF and storing the key file in a different location outside the repo. Then it becomes an issue of safety, people need to store the file in a different location to be able to access the repo in case the machine they're saving data from becomes unaccessible.

Sure, we could just blame the users when something goes wrong and they lose access to their data, but that's not what we will do.

There are two separate components in this issue, and I'm not sure which one you're trying to solve:
 1. The key files are stored outside of the repository
 2. The (expensive) KDF is not needed for accessing the repo

I'm not convinced this is a good feature to add and I'm especially not convinced that this improves security. For restic, we've chosen to trust the algorithms (scrypt, AES, poly1305), if there's an issue we'll see how to mitigate that.

I'm going to close this issue for now, please feel free to add further comments. Thanks!

aawsome commented:
@fd0 It is your project, so it is your decision. I fully understand that you fear it would add a maintenance burden to you and I do appreciate that you would feel responsible for what users would do with restic.
From this point your decision is fully understandable.
At the end, it's all work we are doing in our sparse time and you and the other maintainers so far did an incredible job about restic!

However, I cannot follow your arguments about safety and what should be the user's decision and what restic should ensure. As backup tool, restic is a sysadmin's tool and I would say we should assume that the users do know what they are doing (and yes, to blame them if they use - maybe ignoring a warning - an inappropriate setting).

Actually restic allows passwords like "test" or "password" on one hand and forces users to use long generated passwords (which are usually stored in a password file) if they strongly care about security on the other hand.
Do you feel responsible if some repos saved in an unsafe storage with an unsafe password get leaked? Do you feel responsible if a repo with a generated long password gets inaccessible because the sysadmin only stored a password file which got lost? I hope not!
So I cannot follow your opinion that this optional flag worsens the situation for users, sorry.

>     1. The key files are stored outside of the repository
>     2. The (expensive) KDF is not needed for accessing the repo

As you correctly pointed out, this PR does both. I would like to have a solution for 1. The second point is then a nice side-effect when using this PR.
Would you prefer having a solution where the same key file actually stored in the repository can also be stored only locally? Then I can provide a PR for this. I think it shouldn't be much more complicated than this PR.

For me your actual decision means that I'm going to use a patched version of restic including this PR or maybe switch in future to another backup tool if a more feature-rich alternative shows up. This makes me a bit sad as I really like restic and would like to support it.

fd0 commented:
Thank you very much for the constructive message and the praise for the project! It's seldom that people react in this way when their feature has just been declined :)

> However, I cannot follow your arguments about safety and what should be the user's decision and what restic should ensure. As backup tool, restic is a sysadmin's tool and I would say we should assume that the users do know what they are doing (and yes, to blame them if they use - maybe ignoring a warning - an inappropriate setting).

Just to give you a bit of background: I've been working as a penetration tester for the last decade, I've seen many cryptographic systems fail, so I'm familiar with how such systems fail in practice. As a side effect, I've developed an opinion on what aspects of a system involving cryptography users should be able to influence directly. So in my opinion, the key material for cryptographic functions like AES and Poly1305 should never be exposed to users.

> Actually restic allows passwords like "test" or "password" on one hand and forces users to use long generated passwords (which are usually stored in a password file) if they strongly care about security on the other hand.

Because I don't trust users to select secure passwords, restic uses a very strong Key-Derivation Function (scrypt). It is used to derive the keys for decrypting the master keys stored in the key file. When a new repository is initialized (or a password is added), restic calibrates scrypt so that each call to the KDF takes at least 500ms and uses at least 60MiB of memory ([source](https://github.com/restic/restic/blob/6ec5dc8016f1fb47c922b0c5fb421fe188cf2fc1/internal/repository/key.go#L50-L56)).

Let's assume attackers got access to the files of a repository, but they don't know the (single) password. They can try to find this using a word-list or brute force, but for each try, they'll need 500ms on a CPU and 60MiB of memory. So even on a bunch of GPUs this severely limits the number of passwords that can be tried (which, incidentally, is what scrypt is optimized for).

I think using this construction saves a lot of users who may be a bit careless with selecting a good password.

aawsome commented:
Thank you for your explanation. I strongly appreciate that you take security issues seriously :+1: 
I'm more from the theoretical background. So I do think about theoretical issues (and hope I can contribute from this point) but do not have as much experience in what really fails in practice..

> Interesting, that's what I guessed. Would you mind elaborating why you want that?

Its is as I told that security is always the result of a chain of methods used. Theoretically it is only as save as the weakest link in the chain.

Personally I am maybe very paranoid about data leakage. I do want to have a remote backup repository as this makes sense for back-uped data but I do not trust at all anyone who stores my backup data.
Hence I do need some kind of encryption I trust in combination with a strong key. About the possibility of loosing my key the best method for me would be to print it out on a piece of paper and store this in some physical safe...

Having restic saving the key in the repo adds - as you wrote - scrypt to the security chain.
This means, my data is safe if 

1. my password is strong enough (and not in a word-list, etc that might be within the checks an attacker can do with reasonable effort)
2. the scrypt algorithm cannot be cracked (meaning allowing to find out the derived key without knowing my password within reasonable effort for an attacker)
3. the AES algorithm cannot be cracked (meaning allowing to encrypt the data without knowing the AES key within reasonable effort for an attacker)

Now 1. is in my hands and being paranoid I'll take care about it. About 3. I made my oppinion that I will trust in AES.
About scrypt I do have to admit that I didn't check the algorithm enough to judge it. However regarding the fact that it is much younger than AES and most likely not that well analyzed (yet) I would claim that the biggest threat in the security chain lies in 2. (Please note, that I don't want to say that I believe scrypt can/will be cracked. I just say I believe - with my actual knowledge - it is more likely for scrypt to be cracked than it is for AES.)

So, as a paranoid person, I would like to eliminate risks and therefore use an option which mitigates the risk of scrypt being cracked. This is the case if the key (in whatever form) is only locally stored. Then 2. is completely eliminated and the risk of data leakage is the risk that AES is cracked.
